### PR TITLE
Migrate non delta dbfs tables using Create Table As Select (CTAS). Convert such tables to Delta tables.

### LINKED ISSUE #1340: [FEATURE]: Add CTAS table migration function
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

Parent issues:
- https://github.com/databrickslabs/ucx/issues/889
- https://github.com/databrickslabs/ucx/issues/332

When migrate managed dbfs root non-delta tables, and certain external tables, we need to run CTAS.

### Proposed Solution

- Create a function that can create a UC table using CTAS.
- The CTAS should preserve the partition columns if any.
- The function should be able to be used in either managed or external table migration. In case, external table migration, it should be able to create the table in a provided table location. The location will be passed in by the migration function who calls this CTAS function.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by FastLee**
Migrate Non-Delta DBFS root tables to Managed Delta tables.
closes #1340
This release enhances migration for non-delta DBFS tables, supporting more table types and configurations. New methods improve CTAS functionality with safer SQL statement generation. Additional table format compatibility, new creation methods, and updated integration tests are included. Grant assignments during migration are now supported.

### CODE REVIEW

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 1]
```diff
@@ -203,6 +210,45 @@ def _migrate_dbfs_root_table(self, src_table: Table, rule: Rule, grants: list[Gr
...
+    def _get_create_in_place_sql(self, src_table: Table, rule: Rule) -> str:
+        create_sql = str(next(self._backend.fetch(src_table.sql_show_create())).get("createtab_stmt"))
+        statements = sqlglot.parse(create_sql, read='databricks')
+        assert len(statements) == 1, 'Expected a single statement'
+        create = statements[0]
+        assert isinstance(create, expressions.Create), 'Expected a CREATE statement'
+        # safely replace current table name with the updated catalog
+        for table_name in create.find_all(expressions.Table):
+            if table_name.db == src_table.database and table_name.name == src_table.name:
+                # See https://github.com/tobymao/sqlglot/issues/3311
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:241]: This comment seems irrelevant copy paste here

**Review by nfx — APPROVED** [review]
Lgtm

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 2]
```diff
@@ -203,6 +210,45 @@ def _migrate_dbfs_root_table(self, src_table: Table, rule: Rule, grants: list[Gr
         self._backend.execute(src_table.sql_alter_from(rule.as_uc_table_key, self._ws.get_workspace_id()))
         return self._migrate_acl(src_table, rule, grants)
 
+    def _migrate_non_sync_table(self, src_table: Table, rule: Rule, grants: list[Grant] | None = None):
+        table_migrate_sql = self._get_create_in_place_sql(src_table, rule)
+        logger.debug(f"Migrating table (No Sync) {src_table.key} to using SQL query: {table_migrate_sql}")
+        self._backend.execute(table_migrate_sql)
+        self._backend.execute(src_table.sql_alter_to(rule.as_uc_table_key))
+        self._backend.execute(src_table.sql_alter_from(rule.as_uc_table_key, self._ws.get_workspace_id()))
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:218]: @FastLee : Why does the src_table get both the upgraded "from" and "to" tbl property? I would expect only the "upgraded to" to be applied for the source table

**Review by JCZuurmond — CHANGES_REQUESTED** [review]
Generally LGTM, have a comment to verify

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py (hunk 2 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:219]: The source table holds the SQL to upgrade both the from and to table.
Nice observation.

**Review by nfx — APPROVED** [review]
lgtm
